File: src/main/java/net/coderextreme/x3d_code/www_web3d_org/x3d/content/examples/Basic/development/LODlevelChangedExample.java
package net.coderextreme.x3d_code.www_web3d_org.x3d.content.examples.Basic.development;
import org.web3d.x3d.jsail.*;
import org.web3d.x3d.jsail.CADGeometry.*;
import org.web3d.x3d.jsail.Core.*;
import org.web3d.x3d.jsail.CubeMapTexturing.*;
import org.web3d.x3d.jsail.DIS.*;
import org.web3d.x3d.jsail.EnvironmentalEffects.*;
import org.web3d.x3d.jsail.EnvironmentalSensor.*;
import org.web3d.x3d.jsail.EventUtilities.*;
import org.web3d.x3d.jsail.Followers.*;
import org.web3d.x3d.jsail.Geometry2D.*;
import org.web3d.x3d.jsail.Geometry3D.*;
import org.web3d.x3d.jsail.Geospatial.*;
import org.web3d.x3d.jsail.Grouping.*;
import org.web3d.x3d.jsail.HAnim.*;
import org.web3d.x3d.jsail.Interpolation.*;
import org.web3d.x3d.jsail.KeyDeviceSensor.*;
import org.web3d.x3d.jsail.Layering.*;
import org.web3d.x3d.jsail.Layout.*;
import org.web3d.x3d.jsail.Lighting.*;
import org.web3d.x3d.jsail.NURBS.*;
import org.web3d.x3d.jsail.Navigation.*;
import org.web3d.x3d.jsail.Networking.*;
import org.web3d.x3d.jsail.ParticleSystems.*;
import org.web3d.x3d.jsail.Picking.*;
import org.web3d.x3d.jsail.PointingDeviceSensor.*;
import org.web3d.x3d.jsail.Rendering.*;
import org.web3d.x3d.jsail.RigidBodyPhysics.*;
import org.web3d.x3d.jsail.Scripting.*;
import org.web3d.x3d.jsail.Shaders.*;
import org.web3d.x3d.jsail.Shape.*;
import org.web3d.x3d.jsail.Sound.*;
import org.web3d.x3d.jsail.Text.*;
import org.web3d.x3d.jsail.Texturing3D.*;
import org.web3d.x3d.jsail.Texturing.*;
import org.web3d.x3d.jsail.Time.*;
import org.web3d.x3d.jsail.VolumeRendering.*;
import org.web3d.x3d.jsail.fields.*;
import org.web3d.x3d.sai.*;
import org.web3d.x3d.sai.CADGeometry.*;
import org.web3d.x3d.sai.Core.*;
import org.web3d.x3d.sai.CubeMapTexturing.*;
import org.web3d.x3d.sai.DIS.*;
import org.web3d.x3d.sai.EnvironmentalEffects.*;
import org.web3d.x3d.sai.EnvironmentalSensor.*;
import org.web3d.x3d.sai.EventUtilities.*;
import org.web3d.x3d.sai.Followers.*;
import org.web3d.x3d.sai.Geometry2D.*;
import org.web3d.x3d.sai.Geometry3D.*;
import org.web3d.x3d.sai.Geospatial.*;
import org.web3d.x3d.sai.Grouping.*;
import org.web3d.x3d.sai.HAnim.*;
import org.web3d.x3d.sai.Interpolation.*;
import org.web3d.x3d.sai.KeyDeviceSensor.*;
import org.web3d.x3d.sai.Layering.*;
import org.web3d.x3d.sai.Layout.*;
import org.web3d.x3d.sai.Lighting.*;
import org.web3d.x3d.sai.NURBS.*;
import org.web3d.x3d.sai.Navigation.*;
import org.web3d.x3d.sai.Networking.*;
import org.web3d.x3d.sai.ParticleSystems.*;
import org.web3d.x3d.sai.Picking.*;
import org.web3d.x3d.sai.PointingDeviceSensor.*;
import org.web3d.x3d.sai.Rendering.*;
import org.web3d.x3d.sai.RigidBodyPhysics.*;
import org.web3d.x3d.sai.Scripting.*;
import org.web3d.x3d.sai.Shaders.*;
import org.web3d.x3d.sai.Shape.*;
import org.web3d.x3d.sai.Sound.*;
import org.web3d.x3d.sai.Text.*;
import org.web3d.x3d.sai.Texturing3D.*;
import org.web3d.x3d.sai.Texturing.*;
import org.web3d.x3d.sai.Time.*;
import org.web3d.x3d.sai.VolumeRendering.*;
public class LODlevelChangedExample {
  public static void main(String[] args) {
    ConfigurationProperties.setXsltEngine(ConfigurationProperties.XSLT_ENGINE_NATIVE_JAVA);
    ConfigurationProperties.setDeleteIntermediateFiles(false);
    new LODlevelChangedExample().initialize().toFileJSON("../x3d_code/www_web3d_org/x3d/content/examples/Basic/development/LODlevelChangedExample.new.json");
    }
    public X3DObject initialize() {
      X3DObject X3D0 =  new X3DObject().setProfile("Immersive").setVersion("3.0")
      .setHead(new headObject()
        .addMeta(new metaObject().setName("title").setContent("LODlevelChangedExample.x3d"))
        .addMeta(new metaObject().setName("description").setContent("Example use of (proposed) LOD level_changed outputOnly field to trigger loading of Inline nodes in subsequent levels."))
        .addMeta(new metaObject().setName("creator").setContent("Don Brutzman"))
        .addMeta(new metaObject().setName("motivation").setContent("Chris Thorne"))
        .addMeta(new metaObject().setName("created").setContent("26 May 2004"))
        .addMeta(new metaObject().setName("modified").setContent("20 October 2019"))
        .addMeta(new metaObject().setName("warning").setContent("under development, evaluating a proposed specification change"))
        .addMeta(new metaObject().setName("identifier").setContent("https://www.web3d.org/x3d/content/examples/Basic/development/LODlevelChangedExample.x3d"))
        .addMeta(new metaObject().setName("generator").setContent("X3D-Edit 3.3, https://savage.nps.edu/X3D-Edit"))
        .addMeta(new metaObject().setName("license").setContent("../license.html")))
      .setScene(new SceneObject()
        .addChild(new WorldInfoObject().setTitle("LODlevelChangedExample.x3d"))
        .addChild(new LODObject().setDEF("activeLOD").setRange(new MFFloatObject(new MFFloat0().getArray()))
          .addChild(new ShapeObject().setDEF("ShapeLevel0")
            .setGeometry(new TextObject().setString(new MFStringObject(new MFString1().getArray()))
              .setFontStyle(new FontStyleObject().setJustify(new MFStringObject(new MFString2().getArray())))))
          .addChild(new InlineObject().setDEF("InlineLevel1").setLoad(false))
          .addChild(new InlineObject().setDEF("InlineLevel2").setLoad(false))
          .addComments(new CommentsBlock("etc.")))
        .addChild(new GroupObject().setDEF("PreloadActivationGroup")
          .addChild(new ScriptObject().setDEF("PreloadLevelsScript")
            .addField(new fieldObject().setType("SFInt32").setName("level_changed").setAccessType(fieldObject.ACCESSTYPE_INPUTONLY))
            .addField(new fieldObject().setType("SFBool").setName("activateLevel1").setAccessType(fieldObject.ACCESSTYPE_OUTPUTONLY))
            .addField(new fieldObject().setType("SFBool").setName("activateLevel2").setAccessType(fieldObject.ACCESSTYPE_OUTPUTONLY))
            .setSourceCode("ecmascript:\n"+
"\n"+
"function level_changed (value, timestamp)\n"+
"{\n"+
"	Browser.print ('level=' + value);\n"+
"	if      (value == 1) activateLevel1 = true;\n"+
"	else if (value == 2) activateLevel2 = true;\n"+
"	// etc.\n"+
"}"))
          .addChild(new ROUTEObject().setFromField("level_changed").setFromNode("activeLOD").setToField("level_changed").setToNode("PreloadLevelsScript"))
          .addChild(new ROUTEObject().setFromField("activateLevel1").setFromNode("PreloadLevelsScript").setToField("load").setToNode("InlineLevel1"))
          .addChild(new ROUTEObject().setFromField("activateLevel2").setFromNode("PreloadLevelsScript").setToField("load").setToNode("InlineLevel2")))
        .addChild(new GroupObject().setDEF("ViewpointGroup")
          .addChild(new ViewpointObject().setDescription("level 0").setPosition(new float[] {0f,0f,4f}))
          .addChild(new ViewpointObject().setDescription("level 1").setPosition(new float[] {0f,0f,8f}))
          .addChild(new ViewpointObject().setDescription("level 2").setPosition(new float[] {0f,0f,12f}))))      ;
    return X3D0;
    }
protected class MFFloat0 {
  protected MFFloatObject getArray() {
    return new MFFloatObject(new float[] {6f,10f});
  }
}
protected class MFString1 {
  protected MFStringObject getArray() {
    return new MFStringObject(new java.lang.String[] {"developmental test scene","LODlevelChangedExample"});
  }
}
protected class MFString2 {
  protected MFStringObject getArray() {
    return new MFStringObject(new java.lang.String[] {"MIDDLE","MIDDLE"});
  }
}
}

File: src/main/node/net/x3djsonld/data/c/x3d-code/www.web3d.org/x3d/content/examples/Basic/development/LODlevelChangedExample.js
var java = require('java');
java.asyncOptions = {
  asyncSuffix: undefined,     // Don't generate node-style methods taking callbacks
  syncSuffix: "",              // Sync methods use the base name(!!)
  promiseSuffix: "Promise",   // Generate methods returning promises, using the suffix Promise.
  promisify: require('util').promisify, // Needs Node.js version 8 or greater, see comment below
  ifReadOnlySuffix: "_alt"
};
var autoclass = require('./X3Dautoclass');

// Javadoc annotations follow, see below for source.
/**
 * <p> Example use of (proposed) LOD level_changed outputOnly field to trigger loading of Inline nodes in subsequent levels. </p>
 <p> Related links: <a href="../../../development/LODlevelChangedExample.java">LODlevelChangedExample.java</a> source, <a href="../../../development/LODlevelChangedExampleIndex.html" target="_top">LODlevelChangedExample catalog page</a>, <a href="https://www.web3d.org/x3d/content/examples/X3dResources.html" target="_blank">X3D Resources</a>, <a href="https://www.web3d.org/x3d/content/examples/X3dSceneAuthoringHints.html" target="_blank">X3D Scene Authoring Hints</a> and <a href="https://www.web3d.org/x3d/content/X3dTooltips.html" target="_blank">X3D Tooltips</a>. </p>
	<table style="color:black; border:0px solid; border-spacing:10px 0px;" summary="Scene Metadata">
		<tr style="background-color:silver; border-color:silver;">
			<td style="text-align:center; padding:10px 0px;"><i>meta tags</i></td>
			<td style="text-align:left;   padding:10px 0px;">&nbsp; Document Metadata </td>
		</tr>

		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> title </i> </td>
			<td> <a href="../../../development/LODlevelChangedExample.x3d">LODlevelChangedExample.x3d</a> </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> description </i> </td>
			<td> Example use of (proposed) LOD level_changed outputOnly field to trigger loading of Inline nodes in subsequent levels. </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> creator </i> </td>
			<td> Don Brutzman </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> motivation </i> </td>
			<td> Chris Thorne </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> created </i> </td>
			<td> 26 May 2004 </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> modified </i> </td>
			<td> 20 October 2019 </td>
		</tr>
		<tr style="color:burntorange">
			<td style="text-align:right; vertical-align: text-top;"> <i> warning </i> </td>
			<td> under development, evaluating a proposed specification change </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> identifier </i> </td>
			<td> <a href="https://www.web3d.org/x3d/content/examples/Basic/development/LODlevelChangedExample.x3d" target="_blank">https://www.web3d.org/x3d/content/examples/Basic/development/LODlevelChangedExample.x3d</a> </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> generator </i> </td>
			<td> X3D-Edit 3.3, <a href="https://savage.nps.edu/X3D-Edit" target="_blank">https://savage.nps.edu/X3D-Edit</a> </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> license </i> </td>
			<td> <a href="../../../development/../license.html">../license.html</a> </td>
		</tr>
		<tr style="background-color:silver; border-color:silver;">
			<td style="text-align:center;" colspan="2">  &nbsp; </td>
		</tr>
	</table>

	<p>
		This program uses the
		<a href="https://www.web3d.org/specifications/java/X3DJSAIL.html" target="_blank">X3D Java Scene Access Interface Library (X3DJSAIL)</a>.
		It has been produced using the 
		<a href="https://www.web3d.org/x3d/stylesheets/X3dToNodeJS.xslt" target="_blank">X3dToNodeJS.xslt</a>
		stylesheet to create NodeJS source code from an <code>.x3d</code> scene.
	</p>

	* @author Don Brutzman
 */

function LODlevelChangedExample
  /** Default constructor to create this object. */
  ()
  {
    /** The initialized model object, created within initialize() method. */
    this.x3dModel = {};
    this.initialize();
    return this;
  }
LODlevelChangedExample.prototype = {
  /** Create and initialize the X3D model. */
  initialize : function ()
  {
  this.x3dModel = (new autoclass.X3D()).setProfile("Immersive").setVersion("3.0")
  .setHead((new autoclass.head())
    .addMeta((new autoclass.meta()).setName("title").setContent("LODlevelChangedExample.x3d"))
    .addMeta((new autoclass.meta()).setName("description").setContent("Example use of (proposed) LOD level_changed outputOnly field to trigger loading of Inline nodes in subsequent levels."))
    .addMeta((new autoclass.meta()).setName("creator").setContent("Don Brutzman"))
    .addMeta((new autoclass.meta()).setName("motivation").setContent("Chris Thorne"))
    .addMeta((new autoclass.meta()).setName("created").setContent("26 May 2004"))
    .addMeta((new autoclass.meta()).setName("modified").setContent("20 October 2019"))
    .addMeta((new autoclass.meta()).setName("warning").setContent("under development, evaluating a proposed specification change"))
    .addMeta((new autoclass.meta()).setName("identifier").setContent("https://www.web3d.org/x3d/content/examples/Basic/development/LODlevelChangedExample.x3d"))
    .addMeta((new autoclass.meta()).setName("generator").setContent("X3D-Edit 3.3, https://savage.nps.edu/X3D-Edit"))
    .addMeta((new autoclass.meta()).setName("license").setContent("../license.html")))
  .setScene((new autoclass.Scene())
    .addChild((new autoclass.WorldInfo()).setTitle("LODlevelChangedExample.x3d"))
    .addChild((new autoclass.LOD("activeLOD")).setRange(java.newArray("float", [6.0,10.0]))
      .addChild((new autoclass.Shape("ShapeLevel0"))
        .setGeometry((new autoclass.Text()).setString(new autoclass.MFString("\"developmental test scene\" \"LODlevelChangedExample\""))
          .setFontStyle((new autoclass.FontStyle()).setJustify(new autoclass.MFString("\"MIDDLE\" \"MIDDLE\"")))))
      .addChild((new autoclass.Inline("InlineLevel1")).setLoad(false))
      .addChild((new autoclass.Inline("InlineLevel2")).setLoad(false))
      .addComments(" etc. "))
    .addChild((new autoclass.Group("PreloadActivationGroup"))
      .addChild((new autoclass.Script("PreloadLevelsScript")).setSourceCode("\n" + 
"        " + "\n" + 
"ecmascript:" + "\n" + 
"\n" + 
"function level_changed (value, timestamp)" + "\n" + 
"{" + "\n" + 
"	Browser.print ('level=' + value);" + "\n" + 
"	if      (value == 1) activateLevel1 = true;" + "\n" + 
"	else if (value == 2) activateLevel2 = true;" + "\n" + 
"	// etc." + "\n" + 
"}" + "\n")
        .addField((new autoclass.field()).setAccessType("inputOnly").setName("level_changed").setType("SFInt32"))
        .addField((new autoclass.field()).setAccessType("outputOnly").setName("activateLevel1").setType("SFBool"))
        .addField((new autoclass.field()).setAccessType("outputOnly").setName("activateLevel2").setType("SFBool")))
      .addChild((new autoclass.ROUTE()).setFromNode("activeLOD").setFromField("level_changed").setToNode("PreloadLevelsScript").setToField("level_changed"))
      .addChild((new autoclass.ROUTE()).setFromNode("PreloadLevelsScript").setFromField("activateLevel1").setToNode("InlineLevel1").setToField("load"))
      .addChild((new autoclass.ROUTE()).setFromNode("PreloadLevelsScript").setFromField("activateLevel2").setToNode("InlineLevel2").setToField("load")))
    .addChild((new autoclass.Group("ViewpointGroup"))
      .addChild((new autoclass.Viewpoint()).setDescription("level 0").setPosition(0.0,0.0,4.0))
      .addChild((new autoclass.Viewpoint()).setDescription("level 1").setPosition(0.0,0.0,8.0))
      .addChild((new autoclass.Viewpoint()).setDescription("level 2").setPosition(0.0,0.0,12.0))));
  },
  // end of initialize() method


  /** Provide a shallow copy of the X3D model.
   * @return LODlevelChangedExample model
   */
  getX3dModel : function()
  {	  
	  return this.x3dModel;
  },
  
  /** Indicate X3DJSAIL validation results for this X3D model.
   * @return validation results plus exception information, if any
   */
  validateSelf : function()
  {
	var validationResult = "";
	var  exceptionResult = "";
	try
	{
		this.initialize();
		
		if ((this.getX3dModel() == null) || (this.getX3dModel().getHead() == null))
		{
			validationResult = "empty scene, nothing to validate. " + this.x3dModel.validate();
			return validationResult;
		}
		validationResult += this.x3dModel.validate(); // walk entire tree to validate correctness
	}
	catch (e)
	{
		exceptionResult = e; // report exception failures, if any
	}
	if  (exceptionResult === "" && validationResult === "")
	     return "success";
	else
	{
		var returnMessage = "";
		if  (exceptionResult !== "" && validationResult !== "")
			returnMessage += "\n*** ";
		returnMessage += exceptionResult;
		if  (exceptionResult === "" && validationResult !== "")
			returnMessage = "\n" + returnMessage; // skip line before meta tags, etc.
		returnMessage += validationResult;
		return returnMessage;
	}
  },
    /** Default main() method provided for test purposes.
     * @param argv input parameters
     */
    main : function (argv)
    {
		var testObject = new LODlevelChangedExample();
		console.log ("LODlevelChangedExample execution self-validation test results: " + testObject.validateSelf());
	}
}
new LODlevelChangedExample().main();
process.exit(0);